specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1098 @@
+- Checks if the destination of the `CrossMsgMeta` is the current subnet; a subnet higher up in the hierarchy; or a subnet that is lower in the hierarchy.
+    - If the `CrossMsgMeta` points to the current subnet or to some other subnet down the current branch of the hierarchy in its `To`, the `CrossMsgMeta` is stored with the subsequent `BottomUpNonce` in `BottomUpMsgsMeta` to notify the `CrossMsgPool` that the cross-msgs inside the `CrossMsgMeta` need to be conveniently executed (or routed down).
+    - If the `CrossMsgsMeta` points to a subnet that needs to be routed up, `SCA` executes the same logic as when a new bottom-up cross-msg is created in the subnet, but appending the `CrossMsgsMeta` into the `Meta` field of `CrossMsgs`. The corresponding `CrossMsgsMeta` of the current checkpoint is created or updated to include this meta for it to be propagated further up in the next checkpoint. Thus, the `CID` of the new `CrossMsgMeta` for the parent includes a single `CID` that already aggregates a link to the `CrossMsgMeta` of the child with cross-net messages that need to go even upper in the hierarchy.
+
+Validators' `CrossMsgPool`s also listen for new `BottomUpMsgsMeta` being included in their subnet SCA. When a new `CrossMsgsMeta` appears in `BottomUpMsgsMeta` (after the committment of a checkpoint), the `CrossMsgPool` checks if `CrossMsgsMeta.Nonce > AppliedBottomUpNonce` to see if it includes cross-msgs that haven't been executed yet. If this is the case, the `CrossMsgPool`:
+- Gets all `CrossMsgsMeta` in `BottomUpMsgsMeta` with `Nonce > AppliedBottomUpNonce`.
+- Gets `CrossMsgMeta.Cid` and makes a request to the subnet content resolution protocol to resolve the `CrossMsgs` behind that `CrossMsgMeta`. These requests are directed to the subnet in `Source` and they resolve the CID from the subnet's `CrossMsgsRegistry`. 
+    - If the resolved `CrossMsgs` only includes elements in the `Msgs` field, they can be directly proposed in the next block for their execution.
+    - If this is not the case, and `CrossMsgs` includes in its `Meta` field `CrossMsgMetas` from its children, then these `CrossMsgsMeta` need to be resolved recursively until all the `CrossMsgsMeta` have been successfully resolved to their underlying messages.
+- Then, as it happened for top-down messages, when the `CrossMsgPool` has all the bottom-up messages to be applied, it waits for a `FinailityThreshold` before proposing the messages to be sure that the checkpoint commitment can be considered final, and all the resolved cross-net messages are proposed for their inclusion and subsequent execution. 
```

Comment:
I thought the pre-requisite for these bottom-up messages to appear on the radar of the parent SCA is for the SA representing the child chain to forward to it a checkpoint, which was assembled by the child SCA and signed by the validators in the child chain. By definition it should be final. 

What exact finality are we waiting before proposing these messages in a block?

Reply:
```suggestion
- Then, as it happened for top-down messages, when the `CrossMsgPool` has all the bottom-up messages to be applied, it waits for a `FinalityThreshold` before proposing the messages to be sure that the checkpoint commitment can be considered final, and all the resolved cross-net messages are proposed for their inclusion and subsequent execution. 
```